crates/ironrdp-rdpdr-native/src/nix/simple.rs

```diff
@@ -0,0 +1,790 @@
+        Some(parent_pos_for_next) => {
+            let mut find_file_name = None;
+            if req_inner.initial_query > 0 {
+                if req_inner.path.ends_with('*') {
+                    let mut parent = backend.file_base.clone();
+                    let query_path = req_inner.path.replace('\\', "/");
+                    let len = query_path.len();
+                    // path ends with *, so its len > 0
+                    #[allow(clippy::arithmetic_side_effects)]
+                    parent.push_str(&query_path[0..len - 1]);
+                    let dirp = if let Ok(real_parent_c) = CString::new(parent.clone()) {
+                        // SAFETY: it is safe to call this on *nix platform
+                        unsafe { opendir(real_parent_c.as_ptr()) }
+                    } else {
+                        std::ptr::null::<DIR>().cast_mut()
+                    };
+                    if !dirp.is_null() {
+                        loop {
+                            // SAFETY: it is safe to call readdir on *nix platform
+                            let first = unsafe { readdir(dirp) };
+                            if first.is_null() {
+                                break;
+                            }
+                            // SAFETY: it is safe to call this for rdp
+                            let dir_name = unsafe { &(*first).d_name };
+                            // SAFETY: it is safe to call this for rdp
+                            let dir_name_in_rs =
+                                unsafe { CStr::from_ptr(dir_name as *const i8).to_string_lossy().into_owned() };
+                            if dir_name_in_rs == "." || dir_name_in_rs == ".." {
+                                continue;
+                            }
+                            parent.push_str(&dir_name_in_rs);
+                            find_file_name = Some(parent);
+                            break;
+                        }
+                        backend
+                            .file_dir_map
+                            .insert(req_inner.device_io_request.file_id, DirBox { dir: dirp });
+                    }
+                } else {
+                    let mut full_path = backend.file_base.clone();
+                    let query_path = req_inner.path.replace('\\', "/");
+                    full_path.push_str(&query_path);
+                    find_file_name = Some(full_path);
+                }
+                make_query_dir_resp(
+                    find_file_name,
+                    req_inner.device_io_request,
+                    req_inner.file_info_class_lvl,
+                    true,
+                )
+            } else {
+                if let Some(dir_box) = backend.file_dir_map.get_mut(&req_inner.device_io_request.file_id) {
+                    // SAFETY: it is safe to call readdir in multiple threads
+                    let next = unsafe { readdir(dir_box.dir) };
```

Comment:
question: You probably didn’t mean that. `readdir` is [not thread-safe](https://man7.org/linux/man-pages/man3/readdir.3.html).
![image](https://github.com/user-attachments/assets/588c3f67-d837-44ed-a39a-167c8e7e36e4)

But you didn’t implement `Sync` on `DirBox`, so we’re not supposed to be sharing references across several threads.